File: cooperative-savings-system/backend/src/main/java/rw/terimbere/csams/modules/cooperative/entity/Cooperative.java
package rw.terimbere.csams.modules.cooperative.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.UUID;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;
import rw.terimbere.csams.shared.common.entity.SoftDeletableEntity;

@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "cooperatives")
public class Cooperative extends SoftDeletableEntity {

    @Column(name = "name", nullable = false)
    private String name;

    @Column(name = "description", length = 2000)
    private String description;

    @Column(name = "registration_number", length = 128)
    private String registrationNumber;

    @Column(name = "contact_email")
    private String contactEmail;

    @Column(name = "contact_phone", length = 32)
    private String contactPhone;

    @Column(name = "address", length = 512)
    private String address;

    @Builder.Default
    @JdbcTypeCode(SqlTypes.CHAR)
    @Column(name = "currency", nullable = false, length = 3)
    private String currency = "RWF";

    @Builder.Default
    @Column(name = "financial_year_start_month", nullable = false)
    private int financialYearStartMonth = 1;

    @Builder.Default
    @Column(name = "monthly_contribution_amount", nullable = false, precision = 19, scale = 4)
    private BigDecimal monthlyContributionAmount = BigDecimal.ZERO;

    @Builder.Default
    @Column(name = "contribution_due_day", nullable = false)
    private int contributionDueDay = 1;

    @Column(name = "logo_file_key", length = 512)
    private String logoFileKey;

    @Builder.Default
    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 32)
    private CooperativeStatus status = CooperativeStatus.ACTIVE;

    @Column(name = "registration_date")
    private LocalDate registrationDate;

    @Column(name = "created_by")
    private UUID createdBy;
}

File: cooperative-savings-system/frontend/src/shared/types/cooperative.ts
export type CooperativeStatus = 'ACTIVE' | 'INACTIVE' | 'SUSPENDED' | 'ARCHIVED'

export const COOPERATIVE_STATUSES: CooperativeStatus[] = [
  'ACTIVE',
  'INACTIVE',
  'SUSPENDED',
  'ARCHIVED',
]

/** Compact shape from GET /cooperatives/mine (selector). */
export interface CooperativeSummary {
  id: string
  name: string
  status: CooperativeStatus
  currency: string
  logoUrl?: string | null
}

export interface Cooperative {
  id: string
  name: string
  description?: string | null
  registrationNumber?: string | null
  contactEmail?: string | null
  contactPhone?: string | null
  address?: string | null
  currency: string
  financialYearStartMonth: number
  monthlyContributionAmount: string | number
  contributionDueDay: number
  registrationDate?: string | null
  status: CooperativeStatus
  logoUrl?: string | null
  logoFileKey?: string | null
  createdAt?: string
  updatedAt?: string
}

export interface CooperativeCreateRequest {
  name: string
  description?: string
  registrationNumber?: string
  contactEmail?: string
  contactPhone?: string
  address?: string
  currency: string
  financialYearStartMonth: number
  monthlyContributionAmount: string | number
  contributionDueDay: number
  registrationDate?: string
}

export type CooperativeUpdateRequest = CooperativeCreateRequest

export interface CooperativeStatusUpdateRequest {
  status: CooperativeStatus
}

export function mapCooperativeSummary(raw: CooperativeSummary): CooperativeSummary {
  return {
    id: String(raw.id),
    name: raw.name,
    status: raw.status,
    currency: raw.currency || 'RWF',
    logoUrl: raw.logoUrl ?? null,
  }
}
